### PR TITLE
[Relay] First order reverse mode Automatic Differentiation

### PR DESCRIPTION
**Pull request by MarisaKirisame**
This is still WIP - memory planner need to handle recursive tuple, and type inference need to do generalize in order for this to work.

### CODE REVIEW

File: include/tvm/relay/ad.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * In deep learning, the most common way to train a deep neural network is by gradient descend or some of it's variant.
+ * Such optimization method require us to input the gradient of neural network, which can be obtained easily using AD.
+ * In fact, back propagation is essentially reverse-mode automatic differentiation, a kind of AD!
+ */
+
+/*! In relay, automatic differentiation(AD) is a macro,
+ *  that transform closed expr(expr without free variable/free type variable) of type
+ *  (x0, x1, x2, ...) -> Float[] to
+ *  (x0, x1, x2, ...) -> (Float[], (x0, x1,  x2, ...)),
```
Comment (grwlf) [inline include/tvm/relay/ad.h]: Could you please clarify the meaning of first tuple item `Float[]` in gradient type `(x0, x1, x2, ...) -> (Float[], (x0, x1,  x2, ...))`? Is it a result value of the original function?

File: python/tvm/relay/op/_tensor.py [hunk 2]
```diff
@@ -80,3 +81,28 @@ def clip_compute(attrs, inputs, output_type, target):
...
+   # (Lhs, Rhs) -> (Out, (Out) -> (Lhs, Rhs))
+   rhs_type = TypeVar('rhs')
+   lhs_type = TypeVar('lhs')
+   out_type = TypeVar('out')
+   rhs = Var('rhs', rhs_type)
+   lhs = Var('lhs', lhs_type)
+   ret_type = FuncType([out_type], TupleType([lhs_type, rhs_type]))
+   ret_type = TupleType([out_type, ret_type])
+   ret_type = FuncType([lhs_type, rhs_type], ret_type)
+   body = Tuple([const(1.0), const(1.0)])
```
Comment (grwlf) [inline python/tvm/relay/op/_tensor.py]: According to the comment above, we expect a Function returning a tuple of Value and another function, returning a tuple of gradients. It looks like `ret_type` definition does match this comment. But the `body` of function below is defined as a tuple of two constants. How is it possible?

File: include/tvm/relay/ad.h (hunk 1 above)
Comment (grwlf) [inline include/tvm/relay/ad.h]: This comment probably doesn't match the below definition of `add_grad`, where we seem to calculate `(x0, x1, x2, ...) -> (Float[], Float[] ->(x0, x1,  x2, ...))`

File: python/tvm/relay/op/_tensor.py (hunk 2 above)
Comment (jroesch) [reply python/tvm/relay/op/_tensor.py]: This isn't fully done, we are blocked on some issues with type checking.

**Comment by tqchen** [comment]
@MarisaKirisame please update https://github.com/dmlc/tvm/issues/2237 or open a new RFC to discuss the signature of the FGradient function used in first order AD(and hopefully can be used in more advanced AD pass)

**Comment by MarisaKirisame** [comment]
@tqchen can you be more specific? the fgradient signature is included in the rfc already.

> Differentiating programs that include operators requires us to register implementations of gradients for the different operators, which we will include as attributes. More specfically, for an operator f : <x0, x1, x2...> -> y, the gradient of the operator is f : <x0, x1, x2...> -> <y, y -> <x0, x1, x2...>>

File: include/tvm/relay/ad.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * By AD, we roughly mean, given a term which denote some mathematical function,
+ * derive a term which denote the derivative of that mathematical function.
+ * Such a method can be compile-time, which is a macro on completely known function.
+ * Formally speaking, such requirement mean that the input function is a closed expression -
+ * that is, it only refer to local variable that is it's parameter, or defined inside it.
+ * Every top level definition satisfy this criteria.
+ * AD can also be run-time, which mean it is merely a function term of AD : (Float[] -> Float[]) -> (Float[] -> Float[]).
+ * In relay we currently only support compile-time AD, but it should be enough for a lot of use case.
+ *
+ * In deep learning, the most common way to train a deep neural network is by gradient descend or some of it's variant.
```
Comment (masahi) [inline include/tvm/relay/ad.h]: descent

File: include/tvm/relay/ad.h [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+#include <tvm/relay/expr.h>
+#include <tvm/relay/module.h>
+
+namespace tvm {
+namespace relay {
+
+/*! What is automatic differentiation(AD) and why is it important?
+ * By AD, we roughly mean, given a term which denote some mathematical function,
+ * derive a term which denote the derivative of that mathematical function.
```
Comment (masahi) [inline include/tvm/relay/ad.h]: denote -> denotes
this line and the line above

**Comment by tqchen** [comment]
@MarisaKirisame Sorry, I meant the signature of the FGradient in C++

File: src/relay/pass/ad.cc [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  auto ty = t.as<FuncTypeNode>();
+  CHECK(ty) << "input should be a function";
+  return FuncTypeNode::make(ty->arg_types,
+                            TupleTypeNode::make({
+                              ty->ret_type,
+                              TupleTypeNode::make(ty->arg_types)}), {}, {});
+}
+
+//! \brief if the expression is a GlobalVar, transform to it's expression.
+Expr DeGlobal(const Module& mod, const Expr& e) {
```
Comment (slyubomirsky) [inline src/relay/pass/ad.cc]: Are nested globals handled appropriately?

Comment (slyubomirsky) [reply src/relay/pass/ad.cc]: Ah, I see. Perhaps a clarifying comment that the expression is expected to be a function might be useful

File: include/tvm/relay/op_attr_types.h [hunk 6]
```diff
@@ -119,6 +119,15 @@ using FForwardRewrite = runtime::TypedPackedFunc<
...
        const Array<Expr>& new_args,
        const NodeRef& ctx)>;
+
+/*!
+ * \brief Gradient for a specific op.
+ *
+ * \return the gradient for each parameters.
+ */
+using FPrimalGradient = runtime::TypedPackedFunc<tvm::Array<Expr>(const Expr& orig_call,
+                                                                  const Expr& gradient)>;
```
[the remark this answers is not in the document]
Comment (MarisaKirisame) [reply include/tvm/relay/op_attr_types.h]: it is not original gradient
@ZihengJiang

[the remark this answers is not in the document]
Comment (ZihengJiang) [reply include/tvm/relay/op_attr_types.h]: I thought it is gradient from the output of the forward graph?

File: include/tvm/relay/op_attr_types.h [hunk 7]
```diff
@@ -119,6 +119,15 @@ using FForwardRewrite = runtime::TypedPackedFunc<
   Expr(const Call& ref_call,
        const Array<Expr>& new_args,
        const NodeRef& ctx)>;
+
+/*!
+ * \brief Gradient for a specific op.
+ *
+ * \return the gradient for each parameters.
+ */
+using FPrimalGradient = runtime::TypedPackedFunc<tvm::Array<Expr>(const Expr& orig_call,
```
Comment (ZihengJiang) [inline include/tvm/relay/op_attr_types.h]: document the arguments

File: python/tvm/relay/ad.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+# pylint: disable=unidiomatic-typecheck
+"""
+The set of automatic differentiation algorithms in Relay.
+"""
+from . import _ad
+
+def fo_with_gradient(expr, mod=None):
+    return _ad.fo_with_gradient(expr, mod)
+
+def with_gradient(expr, mod=None):
```
Comment (ZihengJiang) [inline python/tvm/relay/ad.py]: Personally prefer `gradient` directly. And can we make the `order` as an argument? like `gradient(expr, mod=None, order=1)`, although I am not sure whether the gradient function with different order can share the same interface

Comment (MarisaKirisame) [reply python/tvm/relay/ad.py]: higher order gradient is blocked until we has reference (which I will pr after type inference).
fo_with_gradient is a method that work only on first order code, and dont use reference, hence the difference.

Comment (jroesch) [reply python/tvm/relay/ad.py]: We can just standardize the interface @MarisaKirisame, @ZihengJiang we could use `gradient` name for the operator, the plan was to expose `relay.op.grad(...)` which is expanded to the gradient during a later pass.

File: python/tvm/relay/ad.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# pylint: disable=no-else-return
+# pylint: disable=unidiomatic-typecheck
+"""
+The set of automatic differentiation algorithms in Relay.
+"""
+from . import _ad
+
+def fo_with_gradient(expr, mod=None):
```
Comment (tqchen) [inline python/tvm/relay/ad.py]: Let us consider rename ad-> gradient, as it is more comment name among frameworks

File: include/tvm/relay/builder.h [hunk 10]
```diff
@@ -0,0 +1,69 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file builder.h
+ * \brief A helper file that make building expr easier.
+ */
```
Comment (tqchen) [inline include/tvm/relay/builder.h]: Move these content to https://github.com/dmlc/tvm/blob/master/src/relay/pass/pattern_util.h, for now let us minimize the include files

**Comment by jroesch** [comment]
@MarisaKirisame can you address this feedback and get @ZihengJiang  to give his sign off?

File: python/tvm/relay/backend/interpreter.py [hunk 11]
```diff
@@ -59,6 +72,16 @@ def asnumpy(self):
...
 
+    def __repr__(self):
+        return repr(self.data)
+
+    def __str__(self):
+        return str(self.data)
+
+    def flatten(self):
+        """Remove all layers of nested tuples."""
+        pass
```
Comment (ZihengJiang) [inline python/tvm/relay/backend/interpreter.py]: why pass

File: python/tvm/relay/op/_tensor.py [hunk 12]
```diff
@@ -80,3 +80,16 @@ def clip_compute(attrs, inputs, output_type, target):
     return [topi.clip(inputs[0], attrs.a_min, attrs.a_max)]
 
 register_schedule("clip", schedule_elemwise)
+
+# TODO: Move to C++.
+def add_grad(orig, grad):
+    from tvm.relay import op
```
Comment (ZihengJiang) [inline python/tvm/relay/op/_tensor.py]: move to file header

File: python/tvm/relay/op/op.py [hunk 13]
```diff
@@ -167,6 +167,22 @@ def register_pattern(op_name, pattern, level=10):
     """
     return register(op_name, "TOpPattern", pattern, level)
 
+def register_gradient(op_name, grad_expr, level=10):
```
Comment (ZihengJiang) [inline python/tvm/relay/op/op.py]: `grad_expr` -> `grad_func` or `fgradient`

File: src/relay/pass/ad.cc [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    CHECK(ret) << "cannot downcast";
+    return *ret;
+  }
+};
+
+using ADValue = std::shared_ptr<ADValueNode>;
+
+/*! \brief AD over a program which generates a tensor output. */
+struct ADTensor : ADValueNode {
+  Expr fwd;
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: prefer `forward` and `reverse` to make things more clear

File: src/relay/pass/ad.cc [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+  mutable Expr rev;  // must be an id to avoid duplication
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: what does this comment mean?

File: src/relay/pass/ad.cc [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+  ADValue VisitExpr_(const OpNode* op) final {
+    Op op_ref = GetRef<Op>(op);
+    CHECK(rev_map.count(op_ref))
+      << op->name << " does not have reverse mode defined";
+    return std::make_shared<ADFunction>([this, op_ref](const std::vector<ADValue>& args) {
+        std::vector<Expr> call_args;
+        for (const ADValue& adval : args) {
+          call_args.push_back(adval->get<ADTensor>().fwd);
+        }
+        auto orig = CallNode::make(op_ref, call_args);
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: How about `attrs` and `type_args`?

File: src/relay/pass/ad.cc [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+      });
+  }
+
+  ADValue VisitExpr_(const VarNode* op) final {
+    Var v = GetRef<Var>(op);
+    return env.at(v);
+  }
+};
+
+Expr FOWithGradient(const Expr& re, const Module& mod) {
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: FirstOrderGradient

File: src/relay/pass/ad.cc [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+  // We will then build a sequence of lets which implement reverse mode.
+  Expr body = LetList::With([&](LetList* ll) {
+    ReverseAD reverse_ad(ll);
+    ADValue rev = reverse_ad(e);
+    std::vector<ADValue> args;
+    for (const auto& p : f->params) {
+      args.push_back(std::make_shared<ADTensor>(ll, p));
+    }
+    auto c = rev->get<ADFunction>().func(args);
+    const auto& TADN = c->get<ADTensor>();
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: what does `TADN` mean

File: src/relay/pass/ad.cc [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+ */
+struct ADFunction : ADValueNode {
+  std::function<ADValue(const std::vector<ADValue>&)> func;
+  explicit ADFunction(const std::function<ADValue(const std::vector<ADValue>&)>& func) :
+    func(func) { }
+};
+
+struct ReverseAD : ExprFunctor<ADValue(const Expr &)> {
+  const OpMap<FPrimalGradient> rev_map = Op::GetAttr<FPrimalGradient>("FPrimalGradient");
+  std::vector<std::function<void(LetList* ll)>> replay_actions;
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: `actions` sounds strange, could you come up with a better name to indicate that the list is used to construct gradient expressions?

File: python/tvm/relay/ad.py (hunk 8 above)
Comment (ZihengJiang) [reply python/tvm/relay/ad.py]: @MarisaKirisame My point is the interface, even without high order gradient implementation now.

File: python/tvm/relay/gradient.py [hunk 20]
```diff
@@ -0,0 +1,12 @@
+# pylint: disable=no-else-return
+# pylint: disable=unidiomatic-typecheck
+"""
+The set of automatic differentiation algorithms in Relay.
+"""
+from . import _gradient
+
+def gradient(expr, mod=None, order=1):
```
Comment (ZihengJiang) [inline python/tvm/relay/gradient.py]: add comments for arguments

File: src/relay/pass/ad.cc [hunk 21]
```diff
@@ -0,0 +1,194 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file ad.cc
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: Change the file name to `gradient.cc`

File: src/relay/pass/ad.cc [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+  for (const auto& p : f->params) {
+    vt.push_back(p->type_annotation);
+  }
+  return FunctionNode::make(f->params,
+                            body,
+                            TupleTypeNode::make({f->ret_type, TupleTypeNode::make({})}),
+                            {});
+}
+
+TVM_REGISTER_API("relay._gradient.first_order_gradient")
```
Comment (ZihengJiang) [inline src/relay/pass/ad.cc]: `relay._ir_pass.first_order_gradient`

File: python/tvm/relay/gradient.py [hunk 23]
```diff
@@ -0,0 +1,25 @@
+# pylint: disable=no-else-return
+# pylint: disable=unidiomatic-typecheck
+"""
+The set of automatic differentiation algorithms in Relay.
+"""
+from . import _gradient
+
+
+def gradient(expr, mod=None):
```
Comment (ZihengJiang) [inline python/tvm/relay/gradient.py]: move to `relay.pass`

File: include/tvm/relay/ad.h [hunk 24]
```diff
@@ -0,0 +1,49 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file ad.h
```
Comment (ZihengJiang) [inline include/tvm/relay/ad.h]: do not expose cpp interface for now, move these comments to .cc file

**Comment by MarisaKirisame** [comment]
@ZihengJiang I address the comment

**Pull request merged**